Speed up package asset resolution

nguerrera opened the pull request:
Raise only package assets that are used to build

The new task (ResolvePackageAssets) unlike the old ResolvePackageDependencies, only raises items that that will actually be consumed by the build. The new
items are also much easier to consume in targets and do not require "joins" to resolve paths, querying via WithMetadataValue, etc.

I'm seeing about 10% speed gain in single MVC project, and 15% in OrchardCore on incremental `dotnet build` (with restore).  The gains when the binary log is enabled are more pronounced: about 25% with 4X improvement in working setmand 2X improvement in .binlog size + time to open in structured log viewer! Not to mention, that the log is much easier to interpret without all of the extra items.

There is a breaking change here: 
- If you use the old items from RunResolvePackageDependencies, then you need to depend-on/invoke that target. The code has not been deleted, it just won't be invoked if you don't use it. This is not a common sceanario.
- There are also intermediate _Xxx ("private") items that have disappeared.
- Duplicate/unnecessary metadata has been trimmed from items (e.g. NuGetPackageId vs PackageName).

cc @dsplaisted @livarcocc @mikeharder

nguerrera committed:
Remove unused generalization of dependency diagnostics

nguerrera committed:
Remove bizarre target returns

Example: `Returns="_ActiveTFMFileDependencies"`

This returns the string `"_ActiveTFMFileDependencies"` not
`@(_ActiveTFMFileDependencies)".

Clearly these targets aren't actually run for their return values.

The useless values are cluttering the binlog and confusing me.

nguerrera committed:
Raise only package assets that are used to build

The new task (ResolvePackageAssets) unlike the old ResolvePackageDependencies,
only raises items that that will actually be consumed by the build. The new
items are also much easier to consume in targets and do not require "joins"
to resolve paths, querying via WithMetadataValue, etc.

mikeharder commented:
I'm seeing approximately the same perf improvements:

Solution|Operation|7947|7947-ResolvePackageAssets|Improvement
-----|-----|-----|-----|-----
WebSmall|BuildIncrementalSourceChangedLeaf|2.8|2.5|13%
WebLarge|BuildIncrementalSourceChangedLeaf|15.2|14.7|3%
OrchardCore|BuildIncrementalNoChange|22.5|18.9|16%

I suspect the improvement for WebLarge is relatively small because most of the projects are class libraries with no package references.  I will add package references to the WebLarge class libraries to make it closer to a real-world solution (like OrchardCore).

nguerrera committed:
Move validation for .NET Core RID in-box

Previously, Microsoft.NETCore.App package would do this in a custom
target with the PackageDependency items that we no longer raise by
default.

Also, use repo-toolset to generate designer code for resx during build.
There was a  project system bug preventing it from getting generated in VS.

nguerrera commented:
Ready for review. I am still looking into adding some end-to-end tests for analyzers and content files because there don't seem to be any and they are impacted by this.

livarcocc commented on src/Tasks/Microsoft.NET.Build.Tasks/FindItemsFromPackages.cs:
```diff
@@ -13,7 +13,7 @@ namespace Microsoft.NET.Build.Tasks
     /// </summary>
     /// <remarks>
-    /// Both Items and Packages are expected to have 'PackageName' and 'PackageVersion' 
+    /// Both Items and Packages are expected to have ('PackageName' and 'PackageVersion)' or ('NuGetPackageId' and 'NuGetPackageVersion')
```
What is the difference between PackageName and NuGetPackageId?

nguerrera replied:
Different conventions were used in different places. Semantically, they are the same.

livarcocc commented on src/Tasks/Microsoft.NET.Build.Tasks/ProduceContentAssets.cs:
```diff
@@ -251,18 +217,17 @@ private void ProduceContentAsset(ITaskItem contentFile)
             }
 
-            if (contentFile.GetBooleanMetadata("copyToOutput") == true)
+            if (contentFile.GetBooleanMetadata("CopyToOutput") == true)
```
Don't you have a MetadataKey for this?

livarcocc commented on src/Tasks/Microsoft.NET.Build.Tasks/ProduceContentAssets.cs:
```diff
@@ -251,18 +217,17 @@ private void ProduceContentAsset(ITaskItem contentFile)
+            if (contentFile.GetBooleanMetadata("CopyToOutput") == true)
             {
-                string outputPath = contentFile.GetMetadata("outputPath");
+                string outputPath = contentFile.GetMetadata("OutputPath");
```
And this.

livarcocc commented on src/Tasks/Microsoft.NET.Build.Tasks/ProduceContentAssets.cs:
```diff
@@ -273,13 +238,12 @@ private void ProduceContentAsset(ITaskItem contentFile)
 
             // TODO if build action is none do we even need to write the processed file above?
-            string buildAction = contentFile.GetMetadata("buildAction");
+            string buildAction = contentFile.GetMetadata("BuildAction");
```
And this.

nguerrera commented:
Adding back WIP because I discovered a regression and don't want this to get merged.

nguerrera committed:
Share helper code between old and new package-resolution tasks

dasMulli commented:
on my system, https://github.com/dasMulli/cli-incremental-perf-testbed  shows ~23% improvement in incremental builds - 24s down to 18.5s. (20 projects referencing `Microsoft.AspNetCore.All` and each other in a chain)

nguerrera commented:
@dasMulli, cool. Just curious, does that test use the binary logger?

nguerrera committed:
Add end-to-end analyzer resolution test

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -36,6 +36,12 @@ public class ResolvePackageAssets : TaskBase
         public bool MarkPackageReferencesAsExternallyResolved { get; set; }
 
+        /// <summary>
+        /// Check that there is at least one package dependency in the RID graph that is not in the 
```
This looks like an incomplete comment.  "Not in the" ...what?

nguerrera replied:
Fixed.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -367,6 +375,17 @@ private string ResolvePackageAssetPath(LockFileTargetLibrary package, string rel
+        {
+            if (EnsureRuntimePackageDependencies && !string.IsNullOrEmpty(RuntimeIdentifier))
+            {
+                if (compileTimeTarget.Libraries.Count >= runtimeTarget.Libraries.Count)
```
I know you discussed this logic with me, but it's probably a good idea to get feedback from the CoreFx and NuGet folks to make sure it will work correctly.

nguerrera replied:
cc @ericstj We discussed this approach.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/LockFileExtensions.cs:
```diff
@@ -97,7 +97,46 @@ public static LockFileTargetLibrary GetLibrary(this LockFileTarget lockFileTarge
+
+        public static HashSet<string> GetProjectFileDependencySet(this LockFile lockFile)
+        {
+            string GetPackageNameFromDependency(string dependency)
```
I'd suggest a comment here explaining what the format of `dependency` is and thus why we're looking for the whitespace.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/LockFileExtensions.cs:
```diff
@@ -169,10 +208,47 @@ public static LockFileTargetLibrary GetLibrary(this LockFileTarget lockFileTarge
+
+
+        // A package is a TransitiveProjectReference if it is a project, is not directly referenced,
+        // and does not contain a placeholder compile time assembly
```
What would cause there to be a project in the assets file that meets the other conditions but has a placeholder file?

nguerrera replied:
I couldn't repro it. I talked to @emgarten and it has something to do with assets flags on project refs, but none that I tried caused it. This is just moved from elsewhere.

dsplaisted reviewed (approved):
I really really like these changes.  It makes the asset resolution much easier to understand, it makes debugging with build logs much easier (since there's not so much junk), and on top of that it improves perf!

I reviewed this commit-by-commit, so some of my comments were addressed in later commits.

💯💯💯💯💯

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -0,0 +1,389 @@
+            ContentFiles = RaisePackageAssets(
+                runtimeTarget,
+                p => p.ContentFiles,
+                filter: asset => !string.IsNullOrEmpty(asset.PPOutputPath),
```
Oops. I forgot to loop back to this. The thing is that we only process the PPOutputPath files in the common case because nuget writes content items directly to generated props otherwise. There's a bool to override it though. I decided to revert the optimization, but now I'm thinking about adding it back. I need to add a content file test still.

nguerrera replied:
Fixed.

nguerrera replied:
Fixed.

nguerrera committed:
Add preprocessed content file end-to-end test

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -0,0 +1,389 @@
+                setup: (asset, item) =>
+                {
+                    string locale = asset.Properties["locale"];
+                    item.SetMetadata(MetadataKeys.Culture, "locale");
```
Fixed.

nguerrera replied:
Fixed.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -0,0 +1,389 @@
+                setup: (asset, item) =>
+                {
+                    string directory = Path.GetDirectoryName(asset.Path);
+                    item.SetMetadata("DestinationSubDirectory", directory + Path.DirectorySeparatorChar);
```
By design. It should start with runtimes. 

This matches https://github.com/dotnet/sdk/blob/b0433decc8278f001c0bbbf053b6d488f51a62b4/src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.ConflictResolution.targets#L48

If you dotnet publish a portable app, the RID specific assets will go to a runtimes/ folder.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -0,0 +1,389 @@
+                    return VB() || !CS();
+
+                default:
+                    Debug.Assert(false, "We should not get this far for a non-VB, non-C# project.");
```
It was wrong. Fixed.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -0,0 +1,389 @@
+            => asset.Path.EndsWith("/_._", StringComparison.Ordinal);
+
+        private static string NormalizeRelativePath(string relativePath)
+            => relativePath.Replace('/', Path.DirectorySeparatorChar);
```
Let's just let corefx do that: https://github.com/dotnet/coreclr/blob/d265eba12fed4ac124ef6b9f3fa06e2dded7c20d/src/mscorlib/src/System/String.Manipulation.cs#L1019

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -0,0 +1,389 @@
+
+        private static string GetPackageNameFromDependency(string dependency)
+        {
+            int indexOfWhiteSpace = IndexOfWhiteSpace(dependency);
```
Fixed
